velox/functions/sparksql/DecimalCompare.cpp

```diff
@@ -115,6 +116,60 @@ class DecimalCompareFunction : public exec::VectorFunction {
       exec::EvalCtx& context,
       VectorPtr& result) const override {
     prepareResults(rows, resultType, context, result);
+    vector_size_t size = rows.end() - rows.begin();
+    if (size > 32) {
+      if (args[0]->isFlatEncoding() && args[1]->isFlatEncoding()) {
+        const __restrict A* rawA =
+            args[0]->asUnchecked<FlatVector<A>>()->template rawValues<A>();
+        const __restrict B* rawB =
+            args[1]->asUnchecked<FlatVector<B>>()->template rawValues<B>();
+        applySimdComparison(
```

Comment:
Do you see improvement for this?  I don't think there is SIMD comparison for int128

Reply:
Yes, there is a benchmark for decimal comparison: velox_sparksql_benchmarks_compare. Below is the result:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/zhli/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/zhli/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"Aptos Narrow", sans-serif;
	mso-font-charset:134;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->
</style>
</head>

<body link="#467886" vlink="#96607D">


Name | Avg Time(us) for main | Avg Time (us) for Auto-SIMD | Gain
-- | -- | -- | --
compare#0%null##gt | 347.24 | 85.68 | 4.052754435
compare#0%null##gt_with_cast | 771.69 | 622.04 | 1.240579384
compare#0%null##gte | 343.21 | 85.08 | 4.03396803
compare#0%null##gte_with_cast | 773.71 | 627.71 | 1.232591483
compare#0%null##lt | 376.76 | 85.35 | 4.414294083
compare#0%null##lt_with_cast | 798.79 | 623.7 | 1.280727914
compare#0%null##lte | 362.7 | 84.91 | 4.271581675
compare#0%null##lte_with_cast | 790.74 | 627.43 | 1.260284016
compare#0%null##eq | 344.22 | 85.48 | 4.026906879
compare#0%null##eq_with_cast | 774.94 | 624.02 | 1.241851223
compare#0%null##neq | 365.99 | 85.28 | 4.29162758
compare#0%null##neq_with_cast | 786.68 | 667.72 | 1.17815851
compare#10%null##gt | 440.69 | 108.84 | 4.048970967
compare#10%null##gt_with_cast | 870.96 | 612.27 | 1.422509677
compare#10%null##gte | 440.55 | 109.22 | 4.033601904
compare#10%null##gte_with_cast | 868.5 | 605.57 | 1.434185974
compare#10%null##lt | 447.36 | 107.95 | 4.144140806
compare#10%null##lt_with_cast | 871.44 | 610.26 | 1.427981516
compare#10%null##lte | 448.95 | 108.75 | 4.128275862
compare#10%null##lte_with_cast | 871.68 | 613.67 | 1.420437695
compare#10%null##eq | 433.08 | 117.5 | 3.685787234
compare#10%null##eq_with_cast | 868.13 | 613.26 | 1.415598604
compare#10%null##neq | 433.98 | 117.33 | 3.698798261
compare#10%null##neq_with_cast | 768.26 | 645.52 | 1.190141281
compare#90%null##gt | 37.71 | 35.22 | 1.070698467
compare#90%null##gt_with_cast | 140.5 | 139.48 | 1.007312876
compare#90%null##gte | 38.37 | 35.24 | 1.088819523
compare#90%null##gte_with_cast | 141 | 138.52 | 1.017903552
compare#90%null##lt | 37.17 | 35.61 | 1.043807919
compare#90%null##lt_with_cast | 138.32 | 138.69 | 0.99733218
compare#90%null##lte | 38.39 | 35.98 | 1.066981656
compare#90%null##lte_with_cast | 139.59 | 137.01 | 1.018830742
compare#90%null##eq | 37.87 | 36.02 | 1.051360355
compare#90%null##eq_with_cast | 139.67 | 138.48 | 1.008593299
compare#90%null##neq | 38.09 | 36.48 | 1.044133772
compare#90%null##neq_with_cast | 175.02 | 173.65 | 1.007889433



</body>

</html>

Reply:
And add a real case from our TPCDS run:
```
PushedFilters: [IsNotNull(ss_quantity), Or(Or(Or(And(And(GreaterThanOrEqual(ss_quantity,0),LessThanOrEqual(ss_quantity,5)),Or(Or(And(GreaterThanOrEqual(ss_list_price,11.00),LessThanOrEqual(ss_list_price,21.00)),And(GreaterThanOrEqual(ss_coupon_amt,460.00),LessThanOrEqual(ss_coupon_amt,1460.00))),And(GreaterThanOrEqual(ss_wholesale_cost,14.00),LessThanOrEqual(ss_wholesale_cost,34.00)))),And(And(GreaterThanOrEqual(ss_quantity,6),LessThanOrEqual(ss_quantity,10)),Or(Or(And(GreaterThanOrEqual(ss_list_price,91.00),LessThanOrEqual(ss_list_price,101.00)),And(GreaterThanOrEqual(ss_coupon_amt,1430.00),LessThanOrEqual(ss_coupon_amt,2430.00))),And(GreaterThanOrEqual(ss_wholesale_cost,32.00),LessThanOrEqual(ss_wholesale_cost,52.00))))),Or(And(And(GreaterThanOrEqual(ss_quantity,11),LessThanOrEqual(ss_quantity,15)),Or(Or(And(GreaterThanOrEqual(ss_list_price,66.00),LessThanOrEqual(ss_list_price,76.00)),And(GreaterThanOrEqual(ss_coupon_amt,920.00),LessThanOrEqual(ss_coupon_amt,1920.00))),And(GreaterThanOrEqual(ss_wholesale_cost,4.00),LessThanOrEqual(ss_wholesale_cost,24.00)))),And(And(GreaterThanOrEqual(ss_quantity,16),LessThanOrEqual(ss_quantity,20)),Or(Or(And(GreaterThanOrEqual(ss_list_price,142.00),LessThanOrEqual(ss_list_price,152.00)),And(GreaterThanOrEqual(ss_coupon_amt,3054.00),LessThanOrEqual(ss_coupon_amt,4054.00))),And(GreaterThanOrEqual(ss_wholesale_cost,80.00),LessThanOrEqual(ss_wholesale_cost,100.00)))))),Or(And(And(GreaterThanOrEqual(ss_quantity,21),LessThanOrEqual(ss_quantity,25)),Or(Or(And(GreaterThanOrEqual(ss_list_price,135.00),LessThanOrEqual(ss_list_price,145.00)),And(GreaterThanOrEqual(ss_coupon_amt,14180.00),LessThanOrEqual(ss_coupon_amt,15180.00))),And(GreaterThanOrEqual(ss_wholesale_cost,38.00),LessThanOrEqual(ss_wholesale_cost,58.00)))),And(And(GreaterThanOrEqual(ss_quantity,26),LessThanOrEqual(ss_quantity,30)),Or(Or(And(GreaterThanOrEqual(ss_list_price,28.00),LessThanOrEqual(ss_list_price,38.00)),And(GreaterThanOrEqual(ss_coupon_amt,2513.00),LessThanOrEqual(ss_coupon_amt,3513.00))),And(GreaterThanOrEqual(ss_wholesale_cost,42.00),LessThanOrEqual(ss_wholesale_cost,62.00))))))]
ReadSchema: struct<ss_quantity:int,ss_wholesale_cost:decimal(7,2),ss_list_price:decimal(7,2),ss_coupon_amt:decimal(7,2)>
```

Before
![image](https://github.com/facebookincubator/velox/assets/10524738/f242eb32-a4d9-4ef3-9837-6a2f391c1cda)

After
![image](https://github.com/facebookincubator/velox/assets/10524738/2ec8ae97-08ae-43c9-b095-453d81cab5d7)

Reply:
Do you know where is the gain coming from?  Probably not SIMD itself.

Reply:
Yes, __restrict doesn't help for comparisons of the int128_t type. However, we split the logic into two parts: comparison and setting results. Most of the gain is related to the second part, where we set the results with a batch size of 32.
```
  const auto vectorEnd = std::max(end - end % 32, vectorBegin);
  for (auto i = vectorBegin; i < vectorEnd; i += 32) {
    auto res = simd::toBitMask(
        xsimd::batch_bool<int8_t>(xsimd::load_unaligned(tempRes + i)));
    uint32_t mask = *(reinterpret_cast<const uint32_t*>(rowsData + i / 8));
    uint32_t* addr = reinterpret_cast<uint32_t*>(rawResult + i / 8);
    *addr = (*addr & ~mask) | (res & mask);
  }
```

Reply:
I changed the logic in `applyAutoSimdComparison` to do 64 comparisons in a batch, please help to check if this makes sense to you.